File: Engine/src/main/java/org/etieskrill/engine/graphics/model/loader/MeshLoader.java
package org.etieskrill.engine.graphics.model.loader;

import org.etieskrill.engine.graphics.gl.VertexArrayObject;
import org.etieskrill.engine.graphics.model.*;
import org.joml.primitives.AABBf;

import java.util.List;

import static org.lwjgl.opengl.GL30C.glBindVertexArray;

public final class MeshLoader {
    //TODO builder
    public static Mesh loadToVAO(List<Vertex> vertices, List<Integer> indices, Material material) {
        return loadToVAO(vertices, indices, material, null, null, null);
    }

    public static Mesh loadToVAO(List<Vertex> vertices, List<Integer> indices, Material material, AABBf boundingBox) {
        return loadToVAO(vertices, indices, material, null, boundingBox, null);
    }

    public static Mesh loadToVAO(List<Vertex> vertices, List<Integer> indices, Material material, Mesh.DrawMode drawMode) {
        return loadToVAO(vertices, indices, material, null, null, drawMode);
    }

    public static Mesh loadToVAO(List<Vertex> vertices, List<Integer> indices, Material material, List<Bone> bones, AABBf boundingBox, Mesh.DrawMode drawMode) {
        VertexArrayObject<Vertex> vao = VertexArrayObject
                .builder(VertexAccessor.getInstance())
                .vertexElements(vertices)
                .indices(indices)
                .build();
        unbindVAO();

        return new Mesh(material, bones != null ? bones : List.of(),
                vao, indices.size(), boundingBox,
                drawMode != null ? drawMode : Mesh.DrawMode.TRIANGLES
        );
    }

    private static void unbindVAO() {
        glBindVertexArray(0);
    }

}

File: Engine/src/jvmMain/java/org/etieskrill/engine/graphics/model/loader/MeshLoader.kt
package org.etieskrill.engine.graphics.model.loader

import org.etieskrill.engine.graphics.gl.VertexArrayObject
import org.etieskrill.engine.graphics.model.Bone
import org.etieskrill.engine.graphics.model.Material
import org.etieskrill.engine.graphics.model.Mesh
import org.etieskrill.engine.graphics.model.Vertex
import org.joml.primitives.AABBf

fun loadToVAO(
    vertices: List<Vertex>,
    indices: List<Int>,
    material: Material,
    bones: List<Bone>? = null,
    boundingBox: AABBf = AABBf(),
    drawMode: Mesh.DrawMode? = null
): Mesh = VertexArrayObject(Vertex.Companion.Accessor, vertices, indices = indices)
    .also { it.unbind() }
    .run { Mesh(material, bones, this, boundingBox, drawMode ?: Mesh.DrawMode.TRIANGLES) }
